Classes/common/Encryption/Keychain/CDTEncryptionKeychainManager.m

```diff
@@ -0,0 +1,193 @@
+//
+
+#import <CommonCrypto/CommonCryptor.h>
+
+#import "CDTEncryptionKeychainManager.h"
+
+#import "CDTEncryptionKeychainUtils.h"
+#import "CDTEncryptionKeychainConstants.h"
+
+#import "CDTLogging.h"
+
+@interface CDTEncryptionKeychainManager ()
+
+@property (strong, nonatomic, readonly) CDTEncryptionKeychainStorage *storage;
+
+@end
+
+@implementation CDTEncryptionKeychainManager
+
+#pragma mark - Init object
+- (instancetype)init { return [self initWithStorage:nil]; }
+
+- (instancetype)initWithStorage:(CDTEncryptionKeychainStorage *)storage
+{
+    self = [super init];
+    if (self) {
+        if (storage) {
+            _storage = storage;
+        } else {
+            CDTLogError(CDTDATASTORE_LOG_CONTEXT, @"Storage is mandatory");
+
+            self = nil;
+        }
+    }
+
+    return self;
+}
+
+#pragma mark - Public methods
+- (NSData *)retrieveEncryptionKeyDataUsingPassword:(NSString *)password
+{
+    CDTEncryptionKeychainData *data = [self.storage encryptionKeyData];
+    if (!data || ![self isValidEncryptionKeyData:data]) {
+        return nil;
+    }
+    
+    NSData *aesKey =
+        [self generateAESKeyUsingPBKDF2ToDerivePassword:password
+                                               withSalt:data.salt
+                                             iterations:data.iterations
+                                                 length:CDTENCRYPTION_KEYCHAIN_AES_KEY_SIZE];
+
+    NSData *dpk = [self decryptCipheredDpk:data.encryptedDPK usingAESWithKey:aesKey iv:data.iv];
+
+    return dpk;
+}
+
+- (NSData *)generateEncryptionKeyDataUsingPassword:(NSString *)password
+{
+    NSData *dpk = nil;
+
+    if (![self encryptionKeyDataAlreadyGenerated]) {
+        dpk = [self generateDpk];
+
+        CDTEncryptionKeychainData *keychainData =
+            [self keychainDataToStoreDpk:dpk encryptedWithPassword:password];
+
+        if (![self.storage saveEncryptionKeyData:keychainData]) {
+            dpk = nil;
+        }
+    }
+
+    return dpk;
+}
+
+- (BOOL)encryptionKeyDataAlreadyGenerated { return [self.storage encryptionKeyDataExists]; }
+
+- (BOOL)clearEncryptionKeyData { return [self.storage clearEncryptionKeyData]; }
+
+#pragma mark - CDTEncryptionKeychainManager+Internal methods
+- (BOOL)isValidEncryptionKeyData:(CDTEncryptionKeychainData *)data
+{
+    // Ensure the num derivations saved, matches what we have
+    if (data.iterations != CDTENCRYPTION_KEYCHAIN_PBKDF2_ITERATIONS) {
+        CDTLogWarn(CDTDATASTORE_LOG_CONTEXT,
+                   @"Number of stored iterations does NOT match the constant value %li",
+                   (long)CDTENCRYPTION_KEYCHAIN_PBKDF2_ITERATIONS);
+
+        return NO;
+    }
+
+    // Ensure IV has the correct length
+    if (data.iv.length != CDTENCRYPTION_KEYCHAIN_AES_IV_SIZE) {
+        CDTLogWarn(CDTDATASTORE_LOG_CONTEXT, @"IV does not have the expected size: %i bytes",
+                   CDTENCRYPTION_KEYCHAIN_AES_IV_SIZE);
+
+        return NO;
+    }
```

Comment:
Why not check against the constants for key size that CCCryptoCreate uses rather than our own in that case? `kCCKeySizeAES128` or whichever?